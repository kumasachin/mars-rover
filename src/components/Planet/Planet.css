* {
    box-sizing: border-box;
}

.control-buttons {
    display: flex;
    gap: 10px;
    margin-bottom: 20px;
}

.control-buttons button {
    padding: 10px 20px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    font-size: 14px;
    font-weight: bold;
}

.control-buttons .init {
    background-color: #28a745;
    color: white;
}

.control-buttons .init:hover {
    background-color: #218838;
}

.control-buttons .config {
    background-color: #007acc;
    color: white;
}

.control-buttons .config:hover {
    background-color: #005a9e;
}

.column {
    float: left;
    width: 50%;
    padding: 10px;
}