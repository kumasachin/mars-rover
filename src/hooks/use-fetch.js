import { useState, useEffect } from "react";

export const useFetch = (url, options, excutionStatus) => {
  const [response, setResponse] = useState(null);
  const [error, setError] = useState(null);
  const fetchData = async () => {
    try {
      const res = await fetch(url, options);
      const json = await res.json();
      setResponse(json);
    } catch (error) {
      setError(error);
    }
  };

  useEffect(() => {
    fetchData();
  }, [excutionStatus]);
  
  return { response, error };
};

export default useFetch;