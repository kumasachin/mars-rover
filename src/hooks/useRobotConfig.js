import { useState, useCallback } from 'react';
import { DEFAULT_MAP_SIZE, DEFAULT_ROBOT } from '../constants';

export const useRobotConfig = (initialConfig) => {
  const [mapConfig, setMapConfig] = useState(
    initialConfig?.map || DEFAULT_MAP_SIZE
  );
  
  const [robots, setRobots] = useState(
    initialConfig?.robots?.map(robot => {
      const [x, y, direction] = robot.currentPosition.split(" ");
      return {
        name: robot.name,
        color: robot.color,
        x: parseInt(x),
        y: parseInt(y),
        direction,
        instructions: robot.instructions
      };
    }) || [DEFAULT_ROBOT]
  );

  const updateMapConfig = useCallback((field, value) => {
    setMapConfig(prev => ({
      ...prev,
      [field]: parseInt(value) || 1
    }));
  }, []);

  const updateRobot = useCallback((index, field, value) => {
    setRobots(prev => {
      const updated = [...prev];
      if (field === 'x' || field === 'y') {
        updated[index][field] = parseInt(value) || 0;
      } else {
        updated[index][field] = value;
      }
      return updated;
    });
  }, []);

  const addRobot = useCallback(() => {
    const newRobot = {
      ...DEFAULT_ROBOT,
      name: `R${robots.length + 1}`,
      color: robots.length % 2 === 0 ? 'green' : 'blue'
    };
    setRobots(prev => [...prev, newRobot]);
  }, [robots.length]);

  const removeRobot = useCallback((index) => {
    if (robots.length > 1) {
      setRobots(prev => prev.filter((_, i) => i !== index));
    }
  }, [robots.length]);

  const getFormattedConfig = useCallback(() => {
    return {
      map: mapConfig,
      lostCell: { x: [], y: [] },
      robots: robots.map(robot => ({
        name: robot.name,
        color: robot.color,
        currentPosition: `${robot.x} ${robot.y} ${robot.direction}`,
        instructions: robot.instructions
      }))
    };
  }, [mapConfig, robots]);

  return {
    mapConfig,
    robots,
    updateMapConfig,
    updateRobot,
    addRobot,
    removeRobot,
    getFormattedConfig
  };
};
