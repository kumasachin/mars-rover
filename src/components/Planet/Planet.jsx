import React, { useState, useEffect } from "react";
import { Grid } from "../Grid/Grid";
import {useFetch} from "../../hooks/use-fetch";
import CONFIG from "../../config/"
import "./Planet.css";

const Planet = ({ name = "unKnown surface", data = null }) => {
  const [robots, setMarsRobot] = useState(data);
  const [excutionStatus, setExcutionStatus] = useState(false);
  const [isDataValid, setDataValidation] = useState(true);
  const { response } = useFetch(CONFIG.PATH.marsapi);
  const onClickHandler = () => {
    setExcutionStatus(true);
  }

  const validateData = (robotdata) => {
    if (
      robotdata.x <= CONFIG.MAX_COORDINATE_LENGTH &&
      robotdata.y <= CONFIG.MAX_COORDINATE_LENGTH &&
      CONFIG.DIRECTIONS.includes(robotdata.d) &&
      robotdata.instructions.replaceAll("L", "").replaceAll("R", "").replaceAll("F", "").length === 0
    ) {
      return true;
    } 
    return false;
  }

  const errorHandler = () => {
    setDataValidation(true);
  }

  useEffect(() => {
    try {
      if (response) {
        const robots = [...response.robots];
        robots.forEach((robot, index) => {
          const position = robot.currentPosition.split(" ");
          robots[index] = {
            ...robot,
            x: parseInt(position[0]),
            y: parseInt(position[1]),
            d: position[2],
          };
          if (!validateData({...robots[index]})) {
            setDataValidation(false);
            return;
          };
        });

        setMarsRobot([...robots]);
      }
    } catch (e) {
      errorHandler();
      console.log("unknow error while rendering planet");
    }
  }, [response]);

  if (!robots) {
    return <div>Loading...</div>;
  }
  
  return (
    <>
      <h2>This is {name}</h2>
      <button className="init" onClick={onClickHandler} type="button">Start Moving Robot</button>
      <div className="row">
        { 
          isDataValid ?
          <Grid
            dimension={response.map}
            robots={robots}
            lostCell={response.lostCell}
            excutionStatus={excutionStatus}
            errorHandler={errorHandler}
          /> : <div>There are some issues in provided data. Please provide correct data</div>
        }
      </div>
    </>
  );
};

export default Planet;
