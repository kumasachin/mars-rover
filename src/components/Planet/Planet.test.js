import { render, screen, cleanup } from '@testing-library/react';
import React from 'react';
import Enzyme, { shallow, configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Planet from './Planet';
import {useFetch} from "../../hooks/use-fetch";

Enzyme.configure({ adapter: new Adapter() })


const data = {
  response: {
    "map": {
      "x": 1611,
      "y": 16
    },
    "lostCell": {
      "x": [

      ],
      "y": [

      ]
    },
    "robots": [
      {
        "name": "Chintu1",
        "color": "green",
        "currentPosition": "0 0 N",
        "x": 0,
        "y": 1,
        "d": "N",
        "instructions": "FFFFFFFFFFFFFFFFFFF"
      },
      {
        "name": "Chintu2",
        "color": "green",
        "currentPosition": "0 14 N",
        "x": 0,
        "y": 1,
        "d": "N",
        "instructions": "FRFFFFFFF"
      }
    ]
  }
};

jest.mock('../../hooks/use-fetch', () => ({
  useFetch: () => (data)
}));
describe("Planet Grid rendering", () => {
  afterEach(cleanup);

  it('should take a snapshot', () => {
    const { asFragment } = render(<Planet />)
    expect(asFragment(<Planet />)).toMatchSnapshot()
  });
});

describe('<Planet /> should pre loading sceent', () => {

  it('should have the button to start robot movement', () => {
    const container = mount(<Planet name="mars" />);

    expect(
      container.html()
    ).toBe("<div>Loading...</div>");
  });

  it('should have the button to start robot movement', async () => {
    const container01 = mount(<Planet name="mars" data={data.response.robots} />);
      expect(
        container01.find("button.init")
      ).toBeDefined()
  });

  it('should have render the robots', async () => {
    const container01 = mount(<Planet name="mars" data={data.response.robots} />);
    container01.find("button.init").simulate("click");
    expect(
        container01.find(".robot").length
      ).toBe(2);
  });

  it('should change ui on click on buttoon', async () => {
    const container01 = mount(<Planet name="mars" data={data.response.robots} />);
    const preClickHtml = container01.html();
    container01.find("button.init").simulate("click");
    expect(container01.html()).toEqual(preClickHtml)
  });
});