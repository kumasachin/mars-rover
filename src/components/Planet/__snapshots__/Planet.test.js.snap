// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Planet Grid rendering should take a snapshot 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;
